### ISSUE TITLE
Feature Request & a few questions

### ISSUE THREAD

**Issue by GitHub-23666** [autonomousvision/projected_gan#45]
Please add a commandline parameter that a new pkl file is saved per snapshot, as is the case in normal stylegan. 

For what are the parameters cbase cmax map-depth? I can enter fantasy values, nothing changes.

I can train kimgs as long as I want with batch-gpu=4, on resume however I ALWAYS get a cuda out of memory? With batch-gpu=2 the resume works then. Why is that so?

Is it possible to train Projected GAN with TPU, if so do you implement this?

I noticed, I specify the learning rate with --dlr=0.0025, but in the training_options.json or directly at the start training it says "lr": 0.0002 is this a bug? 

Are these extreme differences normal compared to stylegan3?
Same dataset with 1 million images, same state of training Kimgs, Projected GAN has 70 million parameters more than the stylegan model, but it looks funny.

I simply added the Generator & Discriminator parameters displayed during training.

Projected GAN Model has: 119.351.119
Stylegan3-T Model has: 47.321.532

Shouldn't Projected GAN be much better?

I'm not a programmer, maybe I'm just misunderstanding something.

Left PG vs Right SG3

https://user-images.githubusercontent.com/29997517/149690336-d323aea1-2bd5-4095-bac2-a27bc0bedc98.mp4

I also noticed that Projected GAN often just produces empty boxes (in the fake images) or really strange patterns compared to Stylegan3.

for example something like this:
![wtf_empty](https://user-images.githubusercontent.com/29997517/149723320-6fafde8d-f792-43be-8e00-e9b0ae927abd.png)
![pg muster](https://user-images.githubusercontent.com/29997517/149722837-2d140f6d-5709-490b-996f-67c88f228238.png)

Collecting environment information...
PyTorch version: 1.10.1
CUDA used to build PyTorch: 11.3
OS: Microsoft Windows 10 Pro

Python version: 3.8.12 (default, Oct 12 2021, 03:01:40) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19043-SP0
Is CUDA available: True
CUDA runtime version: 11.4.120
GPU models and configuration: GPU 0: RTX 2070 Super
Nvidia driver version: 511.09
cuDNN version: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.4\bin\cudnn_ops_train64_8.dll

[training_options.json.txt](https://github.com/autonomousvision/projected_gan/files/7878583/training_options.json.txt)

**Comment by xl-sr** [comment]
> Please add a commandline parameter that a new pkl file is saved per snapshot, as is the case in normal stylegan.

I won't add this feature, since keeping the best and most recent model is enough for most use cases. But feel free to create a fork and add it yourself. It is as simple as adding

```
snapshot_pkl = os.path.join(run_dir, f'network-snapshot-{cur_nimg//1000:06d}.pkl')
if rank == 0:
with open(snapshot_pkl, 'wb') as f:
    dill.dump(snapshot_data, f)
```
after this line: https://github.com/autonomousvision/projected_gan/blob/a9817c8b6722b28f843fa18d605779d706f3e8bb/training/training_loop.py#L391


> For what are the parameters cbase cmax map-depth? I can enter fantasy values, nothing changes.

These are used in the StyleGAN3 repo on our codebase is built. You can either use the different options for configs, change the values directly in the code or add the option yourself.

> I can train kimgs as long as I want with batch-gpu=4, on resume however I ALWAYS get a cuda out of memory? With batch-gpu=2 the resume works then. Why is that so?

Memory issues after restarting is a bug that I will resolve soon.

> Is it possible to train Projected GAN with TPU, if so do you implement this?

I haven't tried it and have no access to a TPU to test it.

> I noticed, I specify the learning rate with --dlr=0.0025, but in the training_options.json or directly at the start training it says "lr": 0.0002 is this a bug?

The lr is coupled to the ```-cfg``` argument, as tuning the lr for FastGAN does not improve results in my experience, hence it is fixed to 0.0002.

>Are these extreme differences normal compared to stylegan3?
>Shouldn't Projected GAN be much better?

Just because the model size is larger it does not mean the model will perform better. The artifacts on face datasets is something we addressed in the paper, I suggest using either ```--cfg=stylegan2``` or to use plain StyleGAN2-ADA for face datasets.

**Comment by GitHub-23666** [comment]
Wow, thats was fast. Thanks.

> > Is it possible to train Projected GAN with TPU, if so do you implement this?
> I haven't tried it and have no access to a TPU to test it.

I only ask because Colab (Free) only has the old K80, but the TPUv2 with 64GB that NOBODY uses is also available, which is equivalent to 4x P100 in terms of performance :) 

> > For what are the parameters cbase cmax map-depth? I can enter fantasy values, nothing changes.
> 
> These are used in the StyleGAN3 repo on our codebase is built. You can either use the different options for configs, change the values directly in the code or add the option yourself.

A suggestion from me, in the train.py --help please highlight this clearly. not everyone who tests your code is a programmer.

how am i supposed to know that some parameters only affect the config stylegan, this is not mentioned anywhere.
a simple only stylegan or fastgan config at train.py --help would not have caused the problem in the first place in my opinion.

It's especially confusing that in your own colab notebook you have specified the paranmeters for fastgan, although they don't seem to have any influence.
I thought I need them also for FastGAN Config.

Quote your own code from the linked Colab Notebook: 

train(
    outdir='training-runs', 
    cfg='fastgan',
    data='./data/art_painting256.zip', 
    gpus=1, 
    batch=64, 
    cond=False, 
    mirror=1, 
    batch_gpu=8, 
    cbase=32768, 
    cmax=512, 
    glr=None, 
    dlr=0.002, 
    desc='', 
    metrics=[],
    kimg=10000, 
    tick=4, 
    snap=1, 
    seed=0, 
    workers=0,
    restart_every=999999,
)

**Comment by xl-sr** [comment]
> It's especially confusing that in your own colab notebook you have specified the paranmeters for fastgan, although they don't seem to have any influence. I thought I need them also for FastGAN Config.

Yes, the colab is circumventing train.py so that is why 

> how am i supposed to know that some parameters only affect the config stylegan, this is not mentioned anywhere.
a simple only stylegan or fastgan config at train.py --help would not have caused the problem in the first place in my opinion.

Yeah, that is a good point, I will add this information :)

**Comment by GitHub-23666** [comment]
Are you sure that the fixed LR of 0.0002 is ok with fastgan? Meanwhile many ticks have passed and it looks much better than before, I wait now once and post new video with 10x more kimgs pg fastgan, which should compensate the LR differences config stylegan3-t 0.0025 to FastGan 0.0002. PG with FastGAN config, is about 3-10x faster than original Stylegan3-t for me, depending on whether local, colab or paid cloud, which should make up for it in the end.

Will read your scientific paper as long as, whereby I will understand probably nothing of it .
Verstehe echt nur "Nur Bahnhof", gibt es das auch auf Deutsch ohne Wissenschaftlichen HG zu haben?
Oder eine AI die arXiv.org Science Papers in Nomalsprache für jeden verständlich übersetzt?
